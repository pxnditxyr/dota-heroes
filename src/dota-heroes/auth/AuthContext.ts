import { createContext } from 'react';

import { User } from '../interfaces/interfaces';

export type AuthContextProps = {
    user: User;
    authLogin: ( user : User ) => void;
    authLogout: ( logged : boolean ) => void;
};

export const AuthContext = createContext<AuthContextProps>( {} as AuthContextProps );
