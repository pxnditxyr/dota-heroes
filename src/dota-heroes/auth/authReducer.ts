import { User } from '../interfaces/interfaces';

type authAction =
    | { type: "authLogin", payload: User }
    | { type: "authLogout", payload: { logged : boolean } }

export const authReducer = ( state : User, action : authAction ) : User => {

    switch ( action.type ) {
        case "authLogin":
            return {
                ...action.payload,
                logged: true,
            }

        case "authLogout":
            return {
                ...state,
                logged: false
            }
        default: 
            return state;
    }
};
