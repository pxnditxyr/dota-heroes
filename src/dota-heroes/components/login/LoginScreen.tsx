import { RouteComponentProps } from 'react-router-dom';

interface Props extends RouteComponentProps<any> {

};

export const LoginScreen : React.FC<Props> = ({ history }) => {

    const handleLogin = () => {
        history.replace( "/" );
    }

    return (
        <>
            <h1> Login </h1>
            <hr />
            <button
                onClick={ handleLogin }
            >
                Login
            </button>
        </>
    );
};
