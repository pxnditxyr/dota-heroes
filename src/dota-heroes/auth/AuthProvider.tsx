import { useEffect, useReducer } from 'react';
import { User } from '../interfaces/interfaces';

import { AuthContext } from './AuthContext';
import { authReducer } from './authReducer';


const INITIAL_STATE : User = {
    id: "",
    name: "",
    email: "",
    password: "",
    logged: false,
};

const init : () => User = () => {
    return JSON.parse( localStorage.getItem( 'user' ) || "{}" );
}

interface AuthProviderProps {
    children: JSX.Element | JSX.Element[];
};

export const AuthProvider = ( { children }: AuthProviderProps ) => {

    const [ user, dispatch ] = useReducer( authReducer, INITIAL_STATE, init );
    useEffect( () => {
        if ( !user ) return;
        localStorage.setItem( 'user', JSON.stringify( user ) );
    }, [ user ] )

    const authLogin = ( user: User ) => {
        dispatch({ type: "authLogin", payload: user });
    };

    const authLogout = ( logged : boolean ) => {
        dispatch({ type: "authLogout", payload: { logged } });
    };
    const auth = () => {
        dispatch({ type: "auth", payload: {} })
    }

    return (
        <AuthContext.Provider
            value={{
                user,
                authLogin,
                authLogout,
                auth,
            }}
        >
            { children }
        </AuthContext.Provider>
    );
};
