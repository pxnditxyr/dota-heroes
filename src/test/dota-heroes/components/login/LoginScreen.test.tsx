import { mount } from 'enzyme';
import { LoginScreen } from '../../../../dota-heroes/components/login/LoginScreen';
import { AuthContext } from '../../../../dota-heroes/auth/AuthContext';
import { MemoryRouter } from 'react-router-dom';

const mockNavigate = jest.fn();

jest.mock( 'react-router-dom', () =>({
    ...( jest.requireActual( 'react-router-dom' ) as object ),
    useNavigate: () => mockNavigate,
}));

describe( "Tests in <LoginScreen />", () => {
    const contextValue = {
        user: {
            id: "",
            name: "",
            email: "",
            password: "",
            logged: false,
        },
        authLogin: jest.fn(),
        authLogout: jest.fn(),
        auth: jest.fn(),
    }
    test( "Must show correctly", () => {
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter initialEntries={[ '/login' ]}>
                    <LoginScreen />
                </MemoryRouter>
            </AuthContext.Provider>
        );
        expect( wrapper ).toMatchSnapshot();
    });

    test( "Must call authLogin with default value, and call navigate", () => {
        const defaultUser = {
            id: "1",
            name: "Yuki",
            email: "pnxditxyr@gmail.com",
            password: "MD5",
            logged: true,
        };
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter initialEntries={[ '/login' ]}>
                    <LoginScreen />
                </MemoryRouter>
            </AuthContext.Provider>
        );
        wrapper.find( 'button' ).simulate( 'click' );

        expect( contextValue.authLogin ).toHaveBeenCalled();
        expect( contextValue.authLogin ).toHaveBeenCalledWith( defaultUser );

        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledWith( '/', { replace: true } );

        localStorage.setItem( "lastPath", '/intelligence' );

        wrapper.find( 'button' ).simulate( 'click' );
        
        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledWith( '/intelligence', { replace: true } );
    });
});
