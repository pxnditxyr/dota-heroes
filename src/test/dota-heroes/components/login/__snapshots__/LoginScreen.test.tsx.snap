// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen /> Must show correctly 1`] = `
Array [
  <h1>
     Login 
  </h1>,
  <hr />,
  <button
    onClick={[Function]}
  >
    Login
  </button>,
]
`;
